### PR TITLE
Fix translation of signed array indices

### LINKED ISSUE #4075: translate-c doesn't cast signed array index
currently (0.5.0+b3dbdf2ad) this
```c
void foo() {
  int a[10], i = 0;
  a[i] = 0;
}
```
gets translated to:
```zig
pub export fn foo() void {
    var a: [10]c_int = undefined;
    var i: c_int = 0;
    a[i] = 0;
}
```
which results in the zig compilation error:
```bash
error: expected type 'usize', found 'c_int'
    a[i] = 0;
      ^
```

I'm not sure exactly what type this should be cast to, but here is one working option, casting to c_uint. 

```diff
diff --git a/test/translate_c.zig b/test/translate_c.zig
index 34d6bb3a0..73e14ee5a 100644
--- a/test/translate_c.zig
+++ b/test/translate_c.zig
@@ -1300,6 +1300,19 @@ pub fn addCases(cases: *tests.TranslateCContext) void {
         \\}
     });
 
+    cases.add("cast signed array index to unsigned",
+        \\void foo() {
+        \\  int a[10], i = 0;
+        \\  a[i] = 0;
+        \\}
+    , &[_][]const u8{
+        \\pub export fn foo() void {
+        \\    var a: [10]c_int = undefined;
+        \\    var i: c_int = 0;
+        \\    a[@intCast(c_uint, i)] = 0;
+        \\}
+    });
+
     cases.add("for loops",
         \\int foo() {
         \\    for (int i = 2, b = 4; i + 2; i = 2) {

```

### PR DESCRIPTION
**Pull request by travisstaloch**
- closes #4075

### CODE REVIEW

**Comment by daurnimator** [comment]
You only need the cast if the index is signed or has more bits than `usize`

**Comment by travisstaloch** [comment]
I thought there might need to be some checks like this.  

Not sure how to do this right now.  I'm guessing the result of ZigClangArraySubscriptExpr_getIdx(stmt) would contain the type info that would need checked. Right?

Here is the line in question:
https://github.com/ziglang/zig/pull/4079/commits/79edfb9d67f8b434120644179715241bbaeea3c9#diff-15590d874e44509c8710e7f4cddb840cR2413

**Comment by andrewrk** [comment]
Thanks!

Now that #4053 is merged, would you mind adding a "Run Translated C" test case?

file: `test/run_translated_c.zig`
cmd: `./zig build test-run-translated-c`

This kind of test is better at catching whether translated-c code will actually turn into viable zig code.

**Comment by travisstaloch** [comment]
@andrewrk  Should the tests be moved to test/run_translated_c.zig and removed from test/translate_c.zig or copied and changed?

**Comment by andrewrk** [comment]
For your pull request, I think it would be nice to copy&change. it's also nice to see the input/output as a test case.

**Comment by travisstaloch** [comment]
I've managed to get this working with the checks for signedness and width. 
However, my solution maybe hacky and wouldn't support casting `long long` indexes.  I found qualTypeIntBitWidth but it doesn't include .LongLong and .ULongLong.  Should these and other c builtin integer types be added here? 

```zig
    switch (ZigClangType_getTypeClass(ty)) {
        .Builtin => {
            const builtin_ty = @ptrCast(*const ZigClangBuiltinType, ty);

            switch (ZigClangBuiltinType_getKind(builtin_ty)) {
                .Char_U,
                .UChar,
                .Char_S,
                .SChar,
                => return 8,
                .UInt128,
                .Int128,
// Should we add more types here such as:
//                .ULongLong,
//                .LongLong,
                => return 128,
                else => return 0,
            }

            unreachable;
        },
```

https://github.com/ziglang/zig/blob/master/src-self-hosted/translate_c.zig#L3043

**Comment by travisstaloch** [comment]
Would I be able to somehow get access to the QualType's TypeInfo?  There seems to be public `.Width` property we could use: https://clang.llvm.org/doxygen/structclang_1_1TypeInfo.html.  That is if we had access to QualType::getTypeInfo().

Or am I looking at the wrong docs here?

**Comment by travisstaloch** [comment]
"You only need the cast if the index is signed or has more bits than usize"
I have no problems checking for signedness.  
@daurnimator @LemonBoy do you guys know how to properly check the width of a type from translate_c.zig?  `qualTypeIntBitWidth()` could work but is likely intentionally incomplete as I noted two comments above.

**Comment by andrewrk** [comment]
Consider that the width of an integer type in bytes depends on the target, and when translating from C to Zig, generally you are translating target-agnostic code to target-agnostic code. So if you have to find out how many bytes an integer is exactly, it's a bit of a code smell. 

There might be a related function that I already made for a different purpose, which selects a type based on another type,  but overrides the signedness. Would that help?

**Comment by travisstaloch** [comment]
I think that would help.  I'm working with a `ZigClangQualType` and trying to deduce the width from that.

**Comment by andrewrk** [comment]
https://github.com/ziglang/zig/blob/8c640b3e604e597ba49abbda39bd97c379072a3a/src-self-hosted/translate_c.zig#L3097-L3105

File: test/translate_c.zig [hunk 1]
```diff
@@ -1764,12 +1764,25 @@ pub fn addCases(cases: *tests.TranslateCContext) void {
         \\pub export var array: [100]c_int = .{0} ** 100;
         \\pub export fn foo(arg_index: c_int) c_int {
         \\    var index = arg_index;
-        \\    return array[index];
+        \\    return array[@intCast(c_uint, index)];
```
Comment (andrewrk) [inline test/translate_c.zig]: Can you add a test case where the index is unsigned, and so it can avoid the `@intCast`?

Comment (travisstaloch) [reply test/translate_c.zig]: :+1:  I already one in my local branch.

**Comment by travisstaloch** [comment]
Thanks for pointing out `transQualTypeIntWidthOf`.  

So rather than checking the width, I'm now planning to create a `narrow128BitIntType` function which just maps  `.LongLong, .ULongLong, .Int128, .UInt128 => c_uint` and leaves other types unchanged.

**Comment by andrewrk** [comment]
Mind rebasing while you're at it? This will make sure your new code doesn't regress the newly added test cases.

**Comment by LemonBoy** [comment]
Thousand dollar question, what about `array[-1]` ? That's legal C and definitely not the same as `array[@bitCast(usize, @intCast(isize, -1))]`.

**Comment by travisstaloch** [comment]
I used to `c_uint` in this PR.  Should these be  `usize` instead?  

Also I used @intCast rather than @bitCast.

**Comment by andrewrk** [comment]
`array[-1]`

```zig
(@as([*]E, &array) - 1).*
```

**Comment by LemonBoy** [comment]
> `(@as([*]E, &array) - 1).*`

That's my point, should this PR generate that if the index is not unsigned?

> Should these be usize instead?

When you're dealing with pointers using a `usize` is recommended.

Also, if you want to keep the PR tidy don't merge `master` into it but use `git rebase` instead.

**Comment by travisstaloch** [comment]
Apologies for the messy commits.  I did rebase but ended up with conflicts and must have taken a wrong turn.  Ended up needing `git rebase --skip` after resolving... :-1: 

> using a usize is recommended

I'll update to `usize` then.

**Comment by travisstaloch** [comment]
I will clean up that nasty merge commit.  Maybe it will be easier to leave as is for now and make a fresh PR if accepted.

**Pull request closed without merging**